backend/server/src/services/tournamentService.ts

```diff
@@ -67,4 +100,262 @@ export class TournamentService {
+    const matches: UnsavedMatch[] = [];
+    for (let i = 0; i < playerIds.length; i++) {
+      for (let j = i + 1; j < playerIds.length; j++) {
+        matches.push({
+          players: [
+            { id: playerIds[i], points: [], color: "red" },
+            { id: playerIds[j], points: [], color: "white" }
+          ],
+          type: "group", // decide what this is going to be.
+          admin: null, // this has to be an ID.
+          elapsedTime: 0,
+          timerStartedTimestamp: null
+          // Set other Match fields as necessary
+        });
+      }
+    }
+    return matches;
+  }
+
+  // private function
+  private async generatePlayoffSchedule(
+    playerIds: Types.ObjectId[],
+    previousMatches: Types.ObjectId[]
+  ): Promise<UnsavedMatch[]> {
+    const matches: UnsavedMatch[] = [];
+    let currentRoundPlayers = [];
+    // check if there's any previous matches.
+    if (previousMatches.length === 0) {
+      // work with original list
+      currentRoundPlayers = [...playerIds];
+    } else {
+      const matchPromises = previousMatches.map(
+        async (matchId) => await MatchModel.findById(matchId).exec()
+      );
+      const matchDatas = await Promise.all(matchPromises);
+
+      for (const matchData of matchDatas) {
+        const winnerId = matchData?.winner;
+        if (winnerId === null || winnerId === undefined) {
+          throw new BadRequestError({
+            message:
+              "Cannot create a new round because the last round of the playoffs is not yet complete."
+          });
+        }
+        currentRoundPlayers.push(winnerId);
+      }
+    }
+
+    for (let i = 0; i < playerIds.length; i += 2) {
+      // only works in power of 2.
+      matches.push({
+        players: [
+          { id: playerIds[i], points: [], color: "red" },
+          { id: playerIds[i + 1], points: [], color: "white" }
+        ],
+        type: "playoff", // decide what this is going to be.
+        admin: null, // this has to be an ID.
+        elapsedTime: 0,
+        timerStartedTimestamp: null
+        // Set other Match fields as necessary
+      });
+    }
+    return matches;
+  }
+
+  // private utility function
+  private isPowerOfTwo(n: number): boolean {
+    if (n <= 0) {
+      return false;
+    }
+    return (n & (n - 1)) === 0;
+  }
+
+  // private utility function
+  private calculateRoundRobinMatches(playerCount: number): number {
+    if (playerCount < 2) {
+      throw new BadRequestError({
+        message:
+          "At least two players are required for a round robin tournament."
+      });
+    }
+    return (playerCount * (playerCount - 1)) / 2;
+  }
+
+  private async tournamentToObject(
+    tournamentObject: Tournament & Document
+  ): Promise<Tournament & Document> {
+    const playerPromises = tournamentObject.players.map(
+      async (playerId) =>
+        await UserModel.findById(playerId)
+          .select("firstName lastName _id")
+          .exec()
+    );
```

Comment:
Population should be used here as well to avoid looping over database requests
```typescript
private async tournamentToObject(
  tournamentObject: Tournament & Document
): Promise<Tournament & Document> {
  const populatedTournament = await TournamentModel.findById(tournamentObject._id)
    .populate({
      path: 'players',
      model: 'User',
      select: 'firstName lastName _id',
    })
    .exec();

  // populatedTournament now contains the players with the specified fields
  return populatedTournament;
}